controllers/usersignup/usersignup_controller.go

```diff
@@ -512,20 +512,32 @@ func (r *Reconciler) updateUserSignupMetricsByState(oldState string, newState st
 }
 
 func (r *Reconciler) generateCompliantUsername(config toolchainconfig.ToolchainConfig, instance *toolchainv1alpha1.UserSignup) (string, error) {
+	// replaced should now be of len
 	replaced := usersignup.TransformUsername(instance.Spec.Username)
-
+	// -4 for "crt-" or "-crt" to be added in following lines. Update this length changing prefix
+	maxlengthWithPrefix := usersignup.MaxLength - 4
 	// Check for any forbidden prefixes
 	for _, prefix := range config.Users().ForbiddenUsernamePrefixes() {
 		if strings.HasPrefix(replaced, prefix) {
-			replaced = fmt.Sprintf("%s%s", "crt-", replaced)
+			if len(replaced) > maxlengthWithPrefix {
+				// replace prefix instead of append
+				replaced = "crt-" + replaced[3:]
```

Comment:
Maybe we should move the forbidden username checks and replacements to the `usersignup.TransformUsername()` function? So we have a single place for all username transformations.

Reply:
+1

Reply:
sure we can definitely move prefix and suffix checks.
But not QualifiedName check, an unqualified name will block UserSignup creation - which I don't think we want to do.

Reply:
done in [047f883](https://github.com/codeready-toolchain/host-operator/pull/794/commits/047f883d16ec5b6326c2d88f295239f994112669)
Also removed the qualifiedName check in [b7cb183](https://github.com/codeready-toolchain/host-operator/pull/794/commits/b7cb18362ad0d592fcc97a1a7e57ae7d78b43df1)as we check for DNSLabel check in toolchain-common